File: exp 15/optimal binary tree.java
import java.util.Scanner;
import java.util.Vector;

public class OBST {

    static void calculateOBST(Vector<Integer> keys, Vector<Double> probabilities) {
        int n = keys.size();
        double[][] cost = new double[n + 1][n + 1];
        double[][] weight = new double[n + 1][n + 1];

        for (int i = 1; i <= n; i++) {
            weight[i][i] = probabilities.get(i - 1);
            cost[i][i] = probabilities.get(i - 1);
        }

        for (int length = 2; length <= n; length++) {
            for (int i = 1; i <= n - length + 1; i++) {
                int j = i + length - 1;
                weight[i][j] = weight[i][j - 1] + probabilities.get(j - 1);
                cost[i][j] = Double.MAX_VALUE;

                for (int r = i; r <= j; r++) {
                    double currentCost = cost[i][r - 1] + cost[r + 1][j] + weight[i][j];
                    if (currentCost < cost[i][j]) {
                        cost[i][j] = currentCost;
                    }
                }
            }
        }
        
        System.out.println("Min cost of OBST: " + cost[1][n]);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Enter the number of keys: ");
        int n = scanner.nextInt();
        
        Vector<Integer> keys = new Vector<>(n);
        Vector<Double> probabilities = new Vector<>(n);
        
        System.out.println("Enter keys: ");
        for (int i = 0; i < n; i++) {
            keys.add(scanner.nextInt());
        }

        System.out.println("Enter probabilities: ");
        for (int i = 0; i < n; i++) {
            probabilities.add(scanner.nextDouble());
        }

        calculateOBST(keys, probabilities);
        scanner.close();
    }
}

File: exp 15/optimal binary tree.cpp
#include <iostream>
#include <vector>
#include <limits>
using namespace std;

void OBST(const vector<int>& keys, const vector<double>& probabilities) {
    int n = keys.size();
    vector<vector<double>> c(n + 1, vector<double>(n + 1, 0.0));
    vector<vector<double>> w(n + 1, vector<double>(n + 1, 0.0));

    for (int i = 1; i <= n; ++i) {
        w[i][i] = probabilities[i - 1];
        c[i][i] = probabilities[i - 1];
    }

    for (int l = 2; l <= n; ++l) {
        for (int i = 1; i <= n - l + 1; ++i) {
            int j = i + l - 1;
            w[i][j] = w[i][j - 1] + probabilities[j - 1];
            c[i][j] = numeric_limits<double>::max();

            for (int r = i; r <= j; ++r) {
                double cur = (r > i ? c[i][r - 1] : 0) + 
                             (r < j ? c[r + 1][j] : 0) + w[i][j];
                if (cur < c[i][j]) {
                    c[i][j] = cur;
                }
            }
        }
    }

    cout << "Min cost of OBST: " << c[1][n] << endl;
}

int main() {
    int n;
    cout << "Enter the number of keys: ";
    cin >> n;

    vector<int> keys(n);
    vector<double> probabilities(n);

    cout << "Enter keys: ";
    for (int i = 0; i < n; ++i) {
        cin >> keys[i];
    }

    cout << "Enter probabilities: ";
    for (int i = 0; i < n; ++i) {
        cin >> probabilities[i];
    }

    OBST(keys, probabilities);

    return 0;
}
